huaweicloud/services/iam/resource_huaweicloud_identity_provider.go

```diff
@@ -0,0 +1,651 @@
+	if err != nil {
+		// If fails to create protocols, then delete the mapping.
+		mErr = multierror.Append(
+			mErr,
+			err,
+			mappings.Delete(client, mappingID),
+		)
+		logp.Printf("[ERROR] Failed to create protocol, so delete the mapping that has been created.\n%s", mErr)
+		return fmtp.Errorf("error in creating provider protocol: %s", mErr.Error())
+	}
+	return nil
+}
+
+// generateLoginLink generate login link base on config.domainID.
+func generateLoginLink(host, domainID, id, protocol string) string {
+	// The domain name is the same as that of the console, it is converted according to the config.Cloud.
+	if host == "myhuaweicloud.com" {
+		host = "huaweicloud.com"
+	}
+	url := fmt.Sprintf("https://auth.%s/authui/federation/websso?domain_id=%s&idp=%s&protocol=%s",
+		host, domainID, id, protocol)
+	return url
+}
+
+func resourceIdentityProviderRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conf := meta.(*config.Config)
+	client, err := conf.IAMNoVersionClient(conf.GetRegion(d))
+	if err != nil {
+		return fmtp.DiagErrorf("Error creating HuaweiCloud IAM client without version number: %s", err)
+	}
+
+	provider, err := providers.Get(client, d.Id())
+	if err != nil {
+		logp.Printf("[ERROR] Error in obtaining identity provider: %s", err)
+		return diag.FromErr(err)
+	}
+
+	// Query the protocol name from HuaweiCloud.
+	protocol := queryProtocolName(client, d)
+	url := generateLoginLink(conf.Cloud, conf.DomainID, provider.ID, protocol)
+
+	status := "disabled"
+	if provider.Enabled {
+		status = "enabled"
+	}
+	mErr := multierror.Append(err,
+		d.Set("name", provider.ID),
+		d.Set("protocol", protocol),
+		d.Set("sso_type", provider.SsoType),
+		d.Set("status", status),
+		d.Set("login_link", url),
+		getAndSetConversionRulesAttrs(client, d),
+		d.Set("description", provider.Description),
+		getAndSetMetadataAttr(client, d),
+		getAndSetAccessTypeAttrs(client, d),
+	)
+
+	if err = mErr.ErrorOrNil(); err != nil {
+		logp.Printf("[ERROR] Error in setting identity provider attributes %s: %s", d.Id(), err)
+		return diag.FromErr(err)
+	}
+	return nil
+}
+
+func getAndSetMetadataAttr(client *golangsdk.ServiceClient, d *schema.ResourceData) error {
+	providerID := d.Get("name").(string)
+	r, err := metadatas.Get(client, providerID, protocolSAML)
+	if err != nil {
+		if errors.As(err, &golangsdk.ErrDefault404{}) {
+			return nil
+		}
+		return err
+	}
+
+	return d.Set("metadata", r.Data)
+}
+
+func getAndSetAccessTypeAttrs(client *golangsdk.ServiceClient, d *schema.ResourceData) error {
+	var mErr *multierror.Error
+	// The access type only work on OIDC protocol.
+	if protocol, ok := d.GetOk("protocol"); ok && protocol == protocolSAML {
+		// Because response_mode and response_type has default values, so set default values for them.
+		mErr = multierror.Append(
+			d.Set("response_mode", "form_post"),
+			d.Set("response_type", "id_token"),
+		)
```

Comment:
why did we set the two fields for saml idp?

Reply:
These two fields have default values, the setting to default values just for import case. 
When importing resources from HuaweiCloud, if we do not set values for them, Terraform will found these arguments have been modified during planning or applying because they have default values.